metricbeat/module/system/raid/blockinfo/parser.go

```diff
@@ -0,0 +1,207 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package blockinfo
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+//get the raid level and use that to determine how we fill out the array
+//Only data-reduntant RIAD levels (1,4,5,6,10) have some of these fields
+func isRedundant(path string) (bool, error) {
+	raw, err := ioutil.ReadFile(filepath.Join(path, "md", "level"))
+	if err != nil {
+		return false, err
+	}
+	raidStr := strings.TrimSpace(string(raw))
+	if raidStr == "raid1" || raidStr == "raid4" || raidStr == "raid5" ||
+		raidStr == "raid6" || raidStr == "raid10" {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func parseIntVal(path string) (int64, error) {
+	var value int64
+	raw, err := ioutil.ReadFile(path)
+	if err != nil {
+		return value, err
+	}
+	strVal := strings.TrimSpace(string(raw))
+
+	value, err = strconv.ParseInt(string(strVal), 10, 64)
+	if err != nil {
+		return value, err
+	}
+
+	return value, nil
+}
+
+//get the current sync status as it exists under md/sync_completed
+//if there's no sync operation in progress, the file will just have 'none'
+//in which case, default to to the overall size
+func getSyncStatus(path string, size int64) (SyncStatus, error) {
+
+	raw, err := ioutil.ReadFile(filepath.Join(path, "md", "sync_completed"))
+	if err != nil {
+		return SyncStatus{}, errors.Wrap(err, "could not open sync_completed")
+	}
+	if string(raw) == "none\n" {
+		return SyncStatus{Complete: size, Total: size}, nil
+	}
+
+	matches := strings.SplitN(strings.TrimSpace(string(raw)), " / ", 2)
+
+	if len(matches) != 2 {
+		return SyncStatus{}, fmt.Errorf("could not get data from sync_completed")
+	}
+
+	current, err := strconv.ParseInt(matches[0], 10, 64)
+	if err != nil {
+		return SyncStatus{}, errors.Wrap(err, "could not parse data sync_completed")
+	}
+
+	total, err := strconv.ParseInt(matches[1], 10, 64)
+	if err != nil {
+		return SyncStatus{}, errors.Wrap(err, "could not parse data sync_completed")
+	}
+
+	return SyncStatus{Complete: current, Total: total}, nil
+
+}
+
+//Create a new disk object, parsing any needed fields
+func newMD(path string) (MDDevice, error) {
+	var dev MDDevice
+
+	dev.Name = filepath.Base(path)
+	size, err := parseIntVal(filepath.Join(path, "size"))
+	if err != nil {
+		return dev, errors.Wrap(err, "could not get device size")
+	}
+	dev.Size = size
+
+	//This is the count of 'active' disks
+	// active, err := parseIntVal(filepath.Join(path, "md", "raid_disks"))
+	// if err != nil {
+	// 	return dev, errors.Wrap(err, "could not get raid_disks")
+	// }
+	// dev.ActiveDisks = active
```

Comment:
Why is this commented out?

Reply:
I moved active disks over to the disk status map, since it seemed more appropriate to have it there.

Reply:
Ok, I see, I guess the resulting number is the same, right?

Reply:
Yep.